### DISCUSSION TITLE
Fragmented handshake messages with CCS between them are not rejected in TLS1.3

### DISCUSSION THREAD

**Discussion by lan1120** [openssl/openssl#25017] [General Discussion]
**Whether other record types contain CCS?**

When OpenSSL receives handshake protocol messages that are fragmented over multiple records split by CCS, the connection is not aborted.

This is a RFC 8446 section 5 violation:
`An implementation may receive an unencrypted record of type change_cipher_spec consisting of the single byte value 0x01 at any time after the first ClientHello message has been sent or received and before the peer's Finished message has been received and MUST  simply drop it without further processing.`

This is a RFC 8446 section 5.1 violation:
`Handshake messages MUST NOT be interleaved with other record types.  That is, if a handshake message is split over two or more records, there MUST NOT be any other records between them.`

**Comment by davidben** [comment]
When we designed the middlebox compatibility logic at the TLS working group, the intention was that CCS handling *not* integrate in the state machine, because some implementations did not want to add extra handshake states. Thus we said the record layer just drops them on the floor without passing to the upper layer.

That means, no, the intention was that receivers do *not* need to drop the connection in this case.

Now, there isn't necessarily a contradiction here because, whether or not the receiver checks it, the sender will never send it. Unfortunately these kinds of MISTs tend to be ambiguous on whether they apply to sender, receiver, or both. This text probably could do with some improvement.
